### PR TITLE
Consolidate snp and tdx crates

### PR DESCRIPTION
**Pull request by mkulke**
# Consolidate az-snp-vtpm and az-tdx-vtpm crates

As planned, since Azure TDX CVMs are in public preview now, this PR consolidates both crates. There is overlap in both TEEs, the common logic has been put into the `hcl` and `vtpm` crates. TDX-specific types have been recreated closer to the public TDX spec w/ less assumptions than in prior revisions in a `tdx` crate. Also added a bit of documentation.

## How to use

```bash
cargo t
cargo doc -p az-snp-vtpm --open
cargo doc -p az-tdx-vtpm --open
```

## Testing done

Ran the test suite, checked the docs.

### CODE REVIEW

File: az-cvm-vtpm/hcl/src/lib.rs [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
-const MAX_REPORT_SIZE: usize = mem::size_of::<SnpVmReport>();
-const MIN_REPORT_SIZE: usize = mem::size_of::<TdxVmReport>();
+const TD_REPORT_SIZE: usize = size_of::<TdReport>();
+const SNP_REPORT_SIZE: usize = size_of::<SnpReport>();
+const MAX_REPORT_SIZE: usize = SNP_REPORT_SIZE;
 const SNP_REPORT_TYPE: u32 = 2;
 const TDX_REPORT_TYPE: u32 = 4;
+const HW_REPORT_OFFSET: usize = offset_of!(AttestationReport, hw_report);
+const fn report_range(report_size: usize) -> RangeInclusive<usize> {
+    HW_REPORT_OFFSET..=(HW_REPORT_OFFSET + report_size)
```
Comment (jepio) [inline az-cvm-vtpm/hcl/src/lib.rs:24]: what is `..=`?

File: az-cvm-vtpm/hcl/src/lib.rs [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
+use std::ops::RangeInclusive;
+use tdx::TdReport;
 use thiserror::Error;
 
 const HCL_AKPUB_KEY_ID: &str = "HCLAkPub";
-const MAX_REPORT_SIZE: usize = mem::size_of::<SnpVmReport>();
-const MIN_REPORT_SIZE: usize = mem::size_of::<TdxVmReport>();
+const TD_REPORT_SIZE: usize = size_of::<TdReport>();
+const SNP_REPORT_SIZE: usize = size_of::<SnpReport>();
+const MAX_REPORT_SIZE: usize = SNP_REPORT_SIZE;
```
Comment (jepio) [inline az-cvm-vtpm/hcl/src/lib.rs:19]: maybe this could be some kind of constexpr `max(SNP_REPORT_SIZE, TD_REPORT_SIZE)`

File: az-cvm-vtpm/hcl/src/lib.rs (hunk 1 above)
Comment (mkulke) [reply az-cvm-vtpm/hcl/src/lib.rs:24]: it's a range expression, albeit the wrong one. good catch.

File: az-cvm-vtpm/hcl/src/lib.rs (hunk 2 above)
Comment (mkulke) [reply az-cvm-vtpm/hcl/src/lib.rs:19]: afaik there is no const max() for iterators, but we can add a simple one.

**Review by jepio — APPROVED** [review]
lgtm based on my reading. maybe @surajssd could also take a look

File: az-cvm-vtpm/az-tdx-vtpm/src/lib.rs [hunk 3]
```diff
@@ -1,38 +1,52 @@
...
 //!  use std::error::Error;
 //!
 //!  fn main() -> Result<(), Box<dyn Error>> {
-//!    let bytes = vtpm::get_report()?;
-//!    let hcl_report = hcl::HclReport::new(bytes)?;
-//!    let tdx_report_slice = hcl_report.tdx_report_slice();
-//!    let report_body = imds::ReportBody::new(tdx_report_slice);
-//!    let td_quote_bytes = imds::get_td_quote(report_body)?;
-//!    let hash = hcl_report.var_data_sha256();
-//!    println!("var_data hash: {:x?}", hash);
```
Comment (iaguis) [inline az-cvm-vtpm/az-tdx-vtpm/src/lib.rs:17]: Why is this removed from the example? Also, shouldn't we check the hash and compare it with the td_report here similar to what we do in SNP?

File: az-cvm-vtpm/az-tdx-vtpm/src/main.rs [hunk 4]
```diff
@@ -2,18 +2,11 @@
...
 use anyhow::Result;
-use az_tdx_vtpm::{hcl, imds, vtpm};
+use az_tdx_vtpm::{imds, report};
 
 fn main() -> Result<()> {
-    let bytes = vtpm::get_report()?;
-    let hcl_report = hcl::HclReport::new(bytes)?;
-
-    let hash = hcl_report.var_data_sha256();
-    println!("var_data hash: {:x?}", hash);
```
Comment (iaguis) [inline az-cvm-vtpm/az-tdx-vtpm/src/main.rs:12]: Same here, why is it removed?

**Review by iaguis — APPROVED** [review]
Nice job removing the duplication and hiding complexity! A nit and a couple of comments. Other than that LGTM.

File: az-cvm-vtpm/az-tdx-vtpm/src/lib.rs (hunk 3 above)
Comment (mkulke) [reply az-cvm-vtpm/az-tdx-vtpm/src/lib.rs:17]: mmh, right, it would be good to add verification sample code to a common example project, but so far the example code only covers SNP so it makes sense to add it to the docs.

File: az-cvm-vtpm/hcl/src/lib.rs [hunk 5]
```diff
@@ -1,21 +1,36 @@
...
+use std::ops::Range;
+use tdx::TdReport;
 use thiserror::Error;
 
 const HCL_AKPUB_KEY_ID: &str = "HCLAkPub";
-const MAX_REPORT_SIZE: usize = mem::size_of::<SnpVmReport>();
-const MIN_REPORT_SIZE: usize = mem::size_of::<TdxVmReport>();
+const TD_REPORT_SIZE: usize = size_of::<TdReport>();
+const SNP_REPORT_SIZE: usize = size_of::<SnpReport>();
+const fn max(a: usize, b: usize) -> usize {
```
[the remark this answers is not in the document]
Comment (mkulke) [reply az-cvm-vtpm/hcl/src/lib.rs]: there is probably a const-max crate or sth :)

File: az-cvm-vtpm/az-tdx-vtpm/src/main.rs (hunk 4 above)
Comment (mkulke) [reply az-cvm-vtpm/az-tdx-vtpm/src/main.rs:12]: readded hash checks
